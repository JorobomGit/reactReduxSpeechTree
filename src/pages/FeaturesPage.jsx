import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import {
  getFeatures, addSentence,
} from '../store/action';
import { ElementList as List } from '../components/ElementList';

function mapStateToProps(state) {
  return {
    features: [...state.features],
  };
}

export class FeaturesPage extends Component {
  constructor() {
    super();
    this.state = {
      contexts: [],
      events: [],
      sentences: [],
      selectedFeatureId: '',
      selectedContextId: '',
      selectedEventId: '',
    };
    this.getContexts = this.getContexts.bind(this);
    this.getEvents = this.getEvents.bind(this);
    this.getSentences = this.getSentences.bind(this);
    this.addSentence = this.addSentence.bind(this);
  }

  componentDidMount() {
    this.props.getFeatures();
  }

  getContexts(selectedFeatureId) {
    const { contexts } = this.props.features.find((feature) => feature.id === selectedFeatureId);
    this.setState({
      contexts: contexts || [], events: [], sentences: [], selectedFeatureId, selectedContextId: '', selectedEventId: '',
    });
  }

  getEvents(selectedContextId) {
    const { events } = this.state.contexts.find((context) => context.id === selectedContextId);
    this.setState({
      events: events || [], sentences: [], selectedContextId, selectedEventId: '',
    });
  }

  getSentences(selectedEventId) {
    const { sentences } = this.state.events.find((event) => event.id === selectedEventId);
    this.setState({ sentences: sentences || [], selectedEventId });
  }

  addSentence(title) {
    const {
      selectedFeatureId, selectedContextId, selectedEventId,
    } = this.state;

    this.props.addSentence({
      selectedFeatureId,
      selectedContextId,
      selectedEventId,
      title,
    });
  }

  render() {
    const {
      contexts, events, sentences,
      selectedEventId, selectedFeatureId, selectedContextId,
    } = this.state;
    return (
      <div>
        <List
          elements={this.props.features}
          clickCallback={this.getContexts}
          selectedItem={selectedFeatureId}
        />
        {selectedFeatureId
        && (
        <List
          elements={contexts}
          clickCallback={this.getEvents}
          selectedItem={selectedContextId}
        />
        ) }
        {selectedContextId
        && (
        <List
          elements={events}
          clickCallback={this.getSentences}
          selectedItem={selectedEventId}
        />
        ) }
        {selectedEventId
        && (
        <List
          elements={sentences}
          addButton={selectedEventId}
          showAddButton={!!selectedEventId}
          addCallback={this.addSentence}
          unselectable
        />
        )}
      </div>
    );
  }
}


FeaturesPage.defaultProps = {
  getFeatures: PropTypes.func,
  addSentence: PropTypes.func,
};

FeaturesPage.propTypes = {
  features: PropTypes.arrayOf(PropTypes.object).isRequired,
  getFeatures: PropTypes.func,
  addSentence: PropTypes.func,
};

const ConnectedFeaturesPage = connect(mapStateToProps, { getFeatures, addSentence })(FeaturesPage);
export default ConnectedFeaturesPage;
