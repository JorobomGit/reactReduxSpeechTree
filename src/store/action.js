// src/js/actions/index.js
import {
  ADD_SENTENCE, FEATURES_LOADED,
} from './constants/action-types';

const addSentence = (payload) => ({ type: ADD_SENTENCE, payload });

const getFeatures = () => (dispatch) => fetch('./mockData/features.json')
  .then((response) => response.json())
  .then((json) => {
    dispatch({ type: FEATURES_LOADED, payload: json });
  });


export {
  addSentence, getFeatures,
};
