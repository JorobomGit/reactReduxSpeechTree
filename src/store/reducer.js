import uuidv4 from 'uuid/v4';
import { ADD_SENTENCE, FEATURES_LOADED } from './constants/action-types';

const initialState = {
  features: [],
};
function rootReducer(state = initialState, action) {
  if (action.type === ADD_SENTENCE) {
    const {
      selectedFeatureId, selectedContextId, selectedEventId, title,
    } = action.payload;
    const featuresCopy = [...state.features];
    const features = featuresCopy.find((feature) => feature.id === selectedFeatureId);
    const contexts = features.contexts.find((context) => context.id === selectedContextId);
    const selectedEvent = contexts.events.find((event) => event.id === selectedEventId);
    const id = uuidv4();
    selectedEvent.sentences.push({
      title,
      id,
    });
    return {
      ...state,
      features: featuresCopy,
    };
  }

  if (action.type === FEATURES_LOADED) {
    return { ...state, features: state.features.concat(action.payload) };
  }
  return state;
}
export default rootReducer;
