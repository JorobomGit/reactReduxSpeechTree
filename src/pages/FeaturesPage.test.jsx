/* eslint-disable max-len */
import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { FeaturesPage } from './FeaturesPage';
import store from '../store/store';
import speechTree from '../../public/mockData/features.json';

configure({ adapter: new Adapter() });
describe('component renders', () => {
  it('without features', () => {
    const getFeatures = jest.fn();
    const addSentence = jest.fn();
    const tree = renderer
      .create(<FeaturesPage store={store} features={[]} getFeatures={getFeatures} addSentence={addSentence} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('with features', () => {
    const getFeatures = jest.fn();
    const addSentence = jest.fn();
    const tree = renderer
      .create(<FeaturesPage store={store} features={speechTree} getFeatures={getFeatures} addSentence={addSentence} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('with features and children', () => {
    const getFeatures = jest.fn();
    const addSentence = jest.fn();
    const wrapper = shallow(<FeaturesPage store={store} features={speechTree} getFeatures={getFeatures} addSentence={addSentence} />);
    const instance = wrapper.instance();

    const featureId = instance.props.features[0].id;
    instance.getContexts(featureId);
    expect(instance.state.contexts).toMatchSnapshot();

    const contextId = instance.state.contexts[0].id;
    instance.getEvents(contextId);
    expect(instance.state.events).toMatchSnapshot();

    const eventId = instance.state.events[0].id;
    instance.getSentences(eventId);
    expect(instance.state.sentences).toMatchSnapshot();

    expect(wrapper.html()).toMatchSnapshot();
  });
});

describe('component behaviour', () => {
  it('addSentence calls redux action', () => {
    const getFeatures = jest.fn();
    const addSentence = jest.fn();
    const wrapper = shallow(
      <FeaturesPage store={store} features={speechTree} getFeatures={getFeatures} addSentence={addSentence} />,
    );
    const instance = wrapper.instance();

    const featureId = instance.props.features[0].id;
    instance.getContexts(featureId);

    const contextId = instance.state.contexts[0].id;
    instance.getEvents(contextId);

    const eventId = instance.state.events[0].id;
    instance.getSentences(eventId);

    instance.addSentence('New sentence');

    expect(addSentence).toHaveBeenCalled();
    expect(wrapper.html()).toMatchSnapshot();
  });
});
