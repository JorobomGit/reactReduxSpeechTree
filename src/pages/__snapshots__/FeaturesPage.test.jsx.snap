// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component behaviour addSentence calls redux action 1`] = `"<div><div><div class=\\"element-list\\"><div class=\\"element-list-card element-list-card-selected\\" role=\\"button\\" tabindex=\\"0\\">Feature 1</div><div class=\\"element-list-card element-list-card-unselected\\" role=\\"button\\" tabindex=\\"0\\">Feature 2</div><div class=\\"element-list-card element-list-card-unselected\\" role=\\"button\\" tabindex=\\"0\\">Feature 3</div></div></div><div><div class=\\"element-list\\"><div class=\\"element-list-card element-list-card-selected\\" role=\\"button\\" tabindex=\\"0\\">Context 1.1</div><div class=\\"element-list-card element-list-card-unselected\\" role=\\"button\\" tabindex=\\"0\\">Context 1.2</div></div></div><div><div class=\\"element-list\\"><div class=\\"element-list-card element-list-card-selected\\" role=\\"button\\" tabindex=\\"0\\">Event 1.1.1</div></div></div><div><div class=\\"element-list\\"><div class=\\"element-list-card element-list-card-unselectable\\" role=\\"button\\" tabindex=\\"0\\">Sentence 1.1.1.1</div></div><div class=\\"element-list-buttons\\"><input type=\\"text\\" value=\\"\\" class=\\"form-control element-list-input\\"/><button disabled=\\"\\" type=\\"button\\" class=\\"btn btn-primary mt-3\\">Add child to this element</button></div></div></div>"`;

exports[`component renders with features 1`] = `
<div>
  <div>
    <div
      className="element-list"
    >
      <div
        className="element-list-card element-list-card-unselected"
        onClick={[Function]}
        onKeyUp={[Function]}
        role="button"
        tabIndex="0"
      >
        Feature 1
      </div>
      <div
        className="element-list-card element-list-card-unselected"
        onClick={[Function]}
        onKeyUp={[Function]}
        role="button"
        tabIndex="0"
      >
        Feature 2
      </div>
      <div
        className="element-list-card element-list-card-unselected"
        onClick={[Function]}
        onKeyUp={[Function]}
        role="button"
        tabIndex="0"
      >
        Feature 3
      </div>
    </div>
    
  </div>
  
  
  
</div>
`;

exports[`component renders with features and children 1`] = `
Array [
  Object {
    "events": Array [
      Object {
        "id": "7b889505-d3c5-4281-ad24-2126c1a7cfdc",
        "sentences": Array [
          Object {
            "id": "122f7f78-14cd-4edf-a8da-6f00588f1a71",
            "title": "Sentence 1.1.1.1",
          },
        ],
        "title": "Event 1.1.1",
      },
    ],
    "id": "1cc34c1f-d440-4a4f-b3e8-a7ec25e7655c",
    "title": "Context 1.1",
  },
  Object {
    "events": Array [],
    "id": "cae5cbe3-75e7-48c5-9c3c-e5ab70cea982",
    "title": "Context 1.2",
  },
]
`;

exports[`component renders with features and children 2`] = `
Array [
  Object {
    "id": "7b889505-d3c5-4281-ad24-2126c1a7cfdc",
    "sentences": Array [
      Object {
        "id": "122f7f78-14cd-4edf-a8da-6f00588f1a71",
        "title": "Sentence 1.1.1.1",
      },
    ],
    "title": "Event 1.1.1",
  },
]
`;

exports[`component renders with features and children 3`] = `
Array [
  Object {
    "id": "122f7f78-14cd-4edf-a8da-6f00588f1a71",
    "title": "Sentence 1.1.1.1",
  },
]
`;

exports[`component renders with features and children 4`] = `"<div><div><div class=\\"element-list\\"><div class=\\"element-list-card element-list-card-selected\\" role=\\"button\\" tabindex=\\"0\\">Feature 1</div><div class=\\"element-list-card element-list-card-unselected\\" role=\\"button\\" tabindex=\\"0\\">Feature 2</div><div class=\\"element-list-card element-list-card-unselected\\" role=\\"button\\" tabindex=\\"0\\">Feature 3</div></div></div><div><div class=\\"element-list\\"><div class=\\"element-list-card element-list-card-selected\\" role=\\"button\\" tabindex=\\"0\\">Context 1.1</div><div class=\\"element-list-card element-list-card-unselected\\" role=\\"button\\" tabindex=\\"0\\">Context 1.2</div></div></div><div><div class=\\"element-list\\"><div class=\\"element-list-card element-list-card-selected\\" role=\\"button\\" tabindex=\\"0\\">Event 1.1.1</div></div></div><div><div class=\\"element-list\\"><div class=\\"element-list-card element-list-card-unselectable\\" role=\\"button\\" tabindex=\\"0\\">Sentence 1.1.1.1</div></div><div class=\\"element-list-buttons\\"><input type=\\"text\\" value=\\"\\" class=\\"form-control element-list-input\\"/><button disabled=\\"\\" type=\\"button\\" class=\\"btn btn-primary mt-3\\">Add child to this element</button></div></div></div>"`;

exports[`component renders without features 1`] = `
<div>
  <div>
    <div
      className="element-list-empty"
    >
      <span>
        This element doesn't have any child yet
      </span>
    </div>
    
  </div>
  
  
  
</div>
`;
